File: R/read.R
#' Read an EAVS data file
#'
#' Reads a single survey year of EAVS data from the cache, downloading it first
#' if necessary. The data are returned as published, with every column as
#' character. Reading everything as text preserves FIPS codes with leading
#' zeros and the survey's missing-value codes without guessing types.
#'
#' Use [eavs_recode_missing()] to turn item columns into numbers, or
#' [eavs_load()] for a cleaned, harmonized panel across years.
#'
#' @param year A single survey year.
#' @param survey `"eavs"` (default) or `"policy"` for the Policy Survey.
#' @param format `"csv"` (default) or `"xlsx"`.
#' @param download Whether to download the file if it is not already cached.
#' @param quiet Suppress progress messages.
#'
#' @return A tibble of the data as published, with `year` and `survey` columns
#'   added at the front. Every other column is character.
#' @export
#'
#' @examples
#' \dontrun{
#' raw_2024 <- eavs_read(2024)
#' }
eavs_read <- function(year, survey = "eavs", format = "csv",
                      download = TRUE, quiet = FALSE) {
  if (length(year) != 1) {
    cli::cli_abort("{.arg year} must be a single year; you gave {length(year)}.")
  }
  survey <- match.arg(survey, c("eavs", "policy"))
  format <- match.arg(format, c("csv", "xlsx"))

  if (download) {
    path <- eavs_download(year, survey = survey, format = format,
                          quiet = quiet)$path
  } else {
    row <- manifest_lookup(year, survey = survey, format = format)
    path <- cache_path(row$file_name)
    if (!file.exists(path)) {
      cli::cli_abort(c(
        "{.val {survey}} {year} ({format}) is not in the cache.",
        i = "Call again with {.code download = TRUE} to fetch it."
      ))
    }
  }

  raw <- switch(format,
    csv  = read_raw_csv(path),
    xlsx = read_raw_xlsx(path)
  )
  tibble::add_column(raw, year = year, survey = survey, .before = 1)
}

read_raw_csv <- function(path) {
  if (identical(tolower(tools::file_ext(path)), "zip")) {
    path <- unzip_member(path, "\\.csv$")
  }
  readr::read_csv(
    path,
    col_types = readr::cols(.default = readr::col_character()),
    locale = readr::locale(encoding = file_encoding(path)),
    show_col_types = FALSE,
    name_repair = "minimal",
    progress = FALSE
  )
}

# EAVS CSVs are usually Windows-1252 (Excel/SPSS exports), which breaks a
# UTF-8 read on place names like "Doña Ana". Detect the encoding: use
# UTF-8 if the bytes are valid UTF-8, otherwise fall back to Windows-1252.
file_encoding <- function(path) {
  s <- tryCatch(
    {
      x <- rawToChar(readBin(path, "raw", n = file.info(path)$size))
      Encoding(x) <- "UTF-8"
      x
    },
    error = function(e) NULL
  )
  if (is.null(s) || !validUTF8(s)) "Windows-1252" else "UTF-8"
}

read_raw_xlsx <- function(path, sheet = 1) {
  require_pkg("readxl", "to read xlsx files")
  if (identical(tolower(tools::file_ext(path)), "zip")) {
    path <- unzip_member(path, "\\.xlsx$")
  }
  readxl::read_excel(
    path,
    sheet = sheet,
    col_types = "text",
    .name_repair = "minimal"
  )
}

# Extract the largest zip member matching `pattern` into a sibling folder in
# the cache and return its path, reusing a previous extraction if present.
unzip_member <- function(zip_path, pattern) {
  contents <- utils::unzip(zip_path, list = TRUE)
  hits <- contents[grepl(pattern, contents$Name, ignore.case = TRUE), , drop = FALSE]
  if (nrow(hits) == 0) {
    cli::cli_abort(
      "No file matching {.val {pattern}} inside {.file {basename(zip_path)}}."
    )
  }
  member <- hits$Name[which.max(hits$Length)]
  exdir <- paste0(zip_path, "-extracted")
  out <- file.path(exdir, member)
  if (!file.exists(out)) {
    utils::unzip(zip_path, files = member, exdir = exdir)
  }
  out
}

File: py/src/tidyeavs/read.py
"""Reading one survey year as published, every column as text."""

from __future__ import annotations

import zipfile
from pathlib import Path

import pandas as pd

from ._constants import FORMATS, SURVEYS
from .cache import cache_path
from .download import download, manifest_lookup


def read(
    year: int,
    survey: str = "eavs",
    format: str = "csv",
    download_if_missing: bool = True,
    quiet: bool = False,
) -> pd.DataFrame:
    """Read a single survey year of EAVS data from the cache.

    Downloads the file first if it is not already cached. The data come back as
    published, with every column read as text. Reading everything as text is
    what keeps FIPS codes' leading zeros and the survey's missing-value codes
    intact instead of guessing types on the way in.

    Use :func:`tidyeavs.recode_missing` to turn item columns into numbers, or
    :func:`tidyeavs.load` for a harmonized panel across years.

    Returns a DataFrame with ``year`` and ``survey`` inserted as the first two
    columns; every other column is a pandas ``string``.
    """
    if survey not in SURVEYS:
        raise ValueError(f"survey must be one of {SURVEYS}; got {survey!r}")
    if format not in FORMATS:
        raise ValueError(f"format must be one of {FORMATS}; got {format!r}")

    if download_if_missing:
        path = Path(download(year, survey=survey, format=format, quiet=quiet)["path"][0])
    else:
        row = manifest_lookup(year, survey=survey, format=format).iloc[0]
        path = cache_path(row["file_name"])
        if not path.is_file():
            raise FileNotFoundError(
                f"{survey} {year} ({format}) is not in the cache. "
                "Call again with download_if_missing=True to fetch it."
            )

    frame = _read_csv(path) if format == "csv" else _read_xlsx(path)
    frame.insert(0, "survey", survey)
    frame.insert(0, "year", year)
    return frame


def _read_csv(path: Path) -> pd.DataFrame:
    if path.suffix.lower() == ".zip":
        path = _extract_member(path, ".csv")
    data = path.read_bytes()
    return pd.read_csv(
        path,
        dtype="string",
        encoding=detect_encoding(data),
        keep_default_na=True,
        na_values=[""],
        low_memory=False,
    )


def _read_xlsx(path: Path, sheet: int = 0) -> pd.DataFrame:
    if path.suffix.lower() == ".zip":
        path = _extract_member(path, ".xlsx")
    return pd.read_excel(path, sheet_name=sheet, dtype="string")


def detect_encoding(data: bytes) -> str:
    """Guess a published EAVS file's encoding from its bytes.

    Some EAVS CSVs are Windows-1252, being Excel and SPSS exports, so a naive
    UTF-8 read fails outright on them. Of the 2016-2024 cycles only 2016 is,
    where the offending bytes are curly quotes in the question-label and comment
    text. Use UTF-8 when the bytes are valid UTF-8 and fall back to Windows-1252
    otherwise, matching ``r/R/read.R``.
    """
    try:
        data.decode("utf-8")
    except UnicodeDecodeError:
        return "cp1252"
    return "utf-8"


def _extract_member(zip_path: Path, suffix: str) -> Path:
    """Extract the largest member with ``suffix`` beside the archive.

    A previous extraction is reused. EAVS zips occasionally carry a small
    readme or codebook alongside the data, so the largest match is the one
    wanted rather than the first.
    """
    with zipfile.ZipFile(zip_path) as archive:
        hits = [
            info
            for info in archive.infolist()
            if info.filename.lower().endswith(suffix) and not info.is_dir()
        ]
        if not hits:
            raise FileNotFoundError(f"No {suffix} file inside {zip_path.name}.")
        member = max(hits, key=lambda info: info.file_size)

        out_dir = zip_path.with_name(zip_path.name + "-extracted")
        out = out_dir / member.filename
        if not out.is_file():
            archive.extract(member, path=out_dir)
    return out
